File: katas/java/src/main/java/mwhelan/Fibonacci.java
package mwhelan;

/**
 * PROBLEM:
 * Given a number n return the Fibonacci number at this
 * position.
 *
 * @author Mike Lowe
 */
public final class Fibonacci {

  // don't want instances
  private Fibonacci() {
  }

  /*
   * Non-recursive solution.
   */

  public static int calculate(final int n) {
    if (n < 0) {
      throw new IllegalArgumentException("n cannot be less than 0 but was " + n);
    }
    if (n < 2) {
      return n;
    }

    int result = 1;
    int prev1 = 1;
    int prev2 = 0;
    for (int j = 2; j <= n; j++) {
      result = prev1 + prev2;
      prev2 = prev1;
      prev1 = result;
    }
    return result;
  }

  /*
   * Recursive solution (not as efficient).
   */

//  public static int calculate(final int n) {
//    if (n < 0) {
//      throw new IllegalArgumentException("n cannot be less than 0 but was " + n);
//    }
//    if (n < 2) {
//      return n;
//    }
//
//    return calculate(n - 1) + calculate(n - 2);
//  }

}

File: katas/javascript/mwhelan/src/Fibonacci.js
/*
 * PROBLEM:
 * Given a number n return the Fibonacci number at this
 * position.
 */

'use strict';

const Fibonacci = {

  calculate: n => {
    if (n == 0) return 0;
    if (n == 1 || n == 2) return 1;
    let current = 2;
    let n1 = 1;
    let n2 = 1;
    for (let i = 3; i <= n; i++) {
      current = n1 + n2;
      n2 = n1;
      n1 = current;
    }
    return current;
  }

  //
  // Recursive solution 
  //
  // calculate: function calculate(n) {
  //   if (n == 0) return 0;
  //   if (n == 1 || n == 2) return 1;
  //   return calculate(n - 1) + calculate(n - 2);
  // }

};

module.exports = Fibonacci;
